File: php-src/MedwatchTable.php
<?php
namespace Maude;

/*
 * Inserts the results of join of the three Maude tables into the medwatch_report table.
 */

class MedwatchTable {

   private $pdo;
   private $select_query;
   private $insert_query;
   private $insert_stmt;

   // bound parameters for prepared insert PDOStatement
   private  $mdr_report_key;
   private  $text_report;
   private  $date_received;
   private  $report_source_code;

   private  $insert_count; 

   public function __construct(\PDO $pdo_in)
   {
           $this->pdo = $pdo_in;
           $this->select_query = <<<SELECT_QUERY
SELECT textfoi.mdr_report_key, textfoi.text_report, mdrfoi.date_received, mdrfoi.report_source_code 
FROM textfoi 
     INNER JOIN
   mdrfoi on textfoi.mdr_report_key=mdrfoi.mdr_report_key
     LEFT OUTER JOIN
   medwatch_report 
     ON textfoi.mdr_report_key=medwatch_report.mdr_report_key
WHERE medwatch_report.mdr_report_key IS NULL ORDER BY textfoi.mdr_report_key
SELECT_QUERY;

         $this->insert_query = <<<INSERT_QUERY
INSERT INTO medwatch_report(mdr_report_key, text_report, date_received, report_source_code) values (:mdr_report_key, :text_report, :date_received,
:report_source_code )
INSERT_QUERY;

         $this->insert_count = 0;
    }
 
    public function insertMaudeData()
    {
       $this->insert_stmt = $this->pdo->prepare($this->insert_query);

       $this->bindInsertParameters($this->insert_stmt); 
       
       $this->insert();  
    }  

    public function getInsertCnt() : int
    {
       return $this->insert_count;
    }

    protected function bindInsertParameters(\PDOStatement $insert_stmt)
    {
       // bind the parameters in each statement
       $insert_stmt->bindParam(':mdr_report_key', $this->mdr_report_key, \PDO::PARAM_INT);
            
       $insert_stmt->bindParam(':text_report', $this->text_report, \PDO::PARAM_STR);
    
       $insert_stmt->bindParam(':date_received', $this->date_received, \PDO::PARAM_STR);
    
       $insert_stmt->bindParam(':report_source_code', $this->report_source_code, \PDO::PARAM_STR);
    } 
    
    protected function insert()
    {
       $this->insert_count = 0;

       $this->pdo->beginTransaction(); 

       try { 

          foreach ($this->pdo->query($this->select_query) as $row) {
       
              $this->mdr_report_key = intval($row['mdr_report_key']);
       
              $this->text_report =  $row['text_report'];
       
              $this->date_received =  $row['date_received'];
       
              $this->report_source_code =  $row['report_source_code'];
       
              $this->insert_stmt->execute();
    
              ++$this->insert_count;
          }

          $this->pdo->commit();

       } catch (\PDOException $e) {
      
           $errors = "\nException Thrown in " . $e->getFile() . " at line " . $e->getLine() . "\n";
              
           $errors .= "Stack Trace as string:\n" . $e->getTraceAsString() . "\n";
                  
           $errors .= "Error message is: " .   $e->getMessage() . "\n";
              
           echo $errors;
       } 
      } 
}

File: src/medwatch-table.cpp
#include "medwatch-table.h"
#include "utility.h"

#include <cppconn/statement.h>
#include <cppconn/resultset.h>
#include <memory>
#include <exception>

using namespace std;
using namespace sql;

// statics
string  medwatch_table::select_query {R"(SELECT textfoi.mdr_rkey, textfoi.text_report, mdrfoi.date_received, mdrfoi.report_source_code 
FROM textfoi INNER JOIN mdrfoi on textfoi.mdr_rkey=mdrfoi.mdr_rkey LEFT OUTER JOIN  medwatch_report  
  ON textfoi.mdr_rkey=medwatch_report.mdr_rkey 
 WHERE medwatch_report.mdr_rkey IS NULL ORDER BY textfoi.mdr_rkey)"};

string medwatch_table::insert_query { "INSERT INTO medwatch_report(mdr_rkey, text_report, date_received, report_source_code, word_cnt) values (?, ?, ?, ?, ?)" };

void medwatch_table::retrieve_and_bind(sql::ResultSet& res) noexcept
{
   // retrieve ResultSet values from LEFT OUTER JOIN...
   mdr_rkey = res.getUInt64(1); 
   
   text_report =  res.getString(2);
   
   date_received =  res.getString(3);
   
   report_source_code = res.getString(4);
       
   // bind them to insert statement
   insert_stmt->setUInt64(1, mdr_rkey); 

   insert_stmt->setString(2, text_report); 

   insert_stmt->setString(3, date_received);

   insert_stmt->setString(4, report_source_code);

   int cnt = word_cnt(text_report);

   insert_stmt->setUInt64(5, cnt); 
} 

void medwatch_table::insertMaudeData() noexcept
{
   insert_count = 0;

   // Run select query
   std::unique_ptr<Statement> select_stmt { conn.createStatement() };
   
   unique_ptr<sql::ResultSet> res { select_stmt->executeQuery(medwatch_table::select_query) };
   
   while (res->next()) {
   
       /* Access column data by numeric offset, 1 is the first column */
       retrieve_and_bind(*res);
 
       insert_stmt->execute();

       ++insert_count;
   }
} 
